src/System.DirectoryServices/tests/System/DirectoryServices/DirectoryServicesTests.cs

```diff
@@ -0,0 +1,147 @@
+
+                    // now let's search for the added data:
+                    SearchOUByName(rootOU, "CoreFxChild1OU");
+                    SearchOUByName(rootOU, "CoreFxChild2OU");
+
+                    SearchCNByName(child1OU, "user.ou1.1");
+                    SearchCNByName(child1OU, "user.ou1.2");
+
+                    SearchCNByName(child2OU, "user.ou2.1");
+                    SearchCNByName(child2OU, "user.ou2.2");
+                }
+                finally
+                {
+                    // rootOU.DeleteTree(); doesn't work as getting "A protocol error occurred. (Exception from HRESULT: 0x80072021)"
+                    DeleteOU(de, "CoreFxRootOU");
+                }
+            }
+        }
+
+        private DirectoryEntry CreateOU(DirectoryEntry de, string ou, string description)
+        {
+            DirectoryEntry ouCoreDevs = de.Children.Add($"ou={ou}","Class");
+            ouCoreDevs.Properties["objectClass"].Value = "organizationalUnit";    // has to be organizationalUnit
+            ouCoreDevs.Properties["description"].Value = description;
+            ouCoreDevs.Properties["ou"].Value = ou;
+            ouCoreDevs.CommitChanges();
+            return ouCoreDevs;
+        }
+
+        private DirectoryEntry CreateCN(DirectoryEntry ouEntry, string cn, string description, string phone)
+        {
+			DirectoryEntry cnEntry = ouEntry.Children.Add($"cn={cn}","Class");
+			cnEntry.Properties["objectClass"].Value = "organizationalRole";
+			cnEntry.Properties["cn"].Value = cn;
+			cnEntry.Properties["description"].Value = description;
+			cnEntry.Properties["ou"].Value = ouEntry.Name;
+			cnEntry.Properties["telephoneNumber"].Value = phone;
+			cnEntry.CommitChanges();
+            return cnEntry;
+        }
+
+        private void DeleteOU(DirectoryEntry parentDe, string ou)
+        {
+            try
+            {
+                // We didn't use DirectoryEntry.DeleteTree as it fails on OpenDJ with "A protocol error occurred. (Exception from HRESULT: 0x80072021)"
+                // Also on AD servers DirectoryEntry.Children.Remove(de) will fail if the de is not a leaf entry. so we had to do it recursively.
+                DirectoryEntry de = parentDe.Children.Find($"ou={ou}");
+                DeleteDirectoryEntry(parentDe, de);
+            }
+            catch
+            {
+                // ignore the error if the test failed early and couldn't create the OU we are trying to delete
+            }
+        }
+
+        private void DeleteDirectoryEntry(DirectoryEntry parent, DirectoryEntry de)
+        {
+			foreach (DirectoryEntry child in de.Children)
+            {
+				DeleteDirectoryEntry(de, child);
+			}
+			
+            parent.Children.Remove(de);
+			parent.CommitChanges();
+        }
+
+        private void SearchOUByName(DirectoryEntry de, string ouName)
+        {
+            using (DirectorySearcher ds = new DirectorySearcher(de))
```

Comment:
Set ds.ClientTimeout to some reasonable value.  I think the default is infinite wait.

Reply:
Thanks. I'll fix that. I'll use 2 minutes timeout which I think can be reasonable.